### PR TITLE
Introduce EndpointResponse<TSuccessContent> and EndpointResponse<TSuccessContent,TErrorContent>

### PR DESCRIPTION
**Pull request by christianhelle**
The changes here introduces versions of `IEndpointResponse` that expose a strongly typed `SuccessContent` and `ErrorContent`. Overloads with generic type parameters in `IMessageResponseBuilder` were implemented for convenience to support the new typed `IEndpointResponse` implementation.

Here's an example usage of `EndpointResponse<TSuccessContent>`

```csharp
public async Task<EndpointResponse<Foo>> GetFooAsync(
    string fooId,
    CancellationToken cancellationToken)
{
    var requestMessage = messageFactory
        .FromTemplate("/v1/foos/{fooId}")
        .WithPathParameter("fooId", fooId)
        .Build(HttpMethod.Get);

    var response = await httpClient.SendAsync(requestMessage, cancellationToken);
    return await messageFactory
        .FromResponse(response)
        .AddSuccessResponse<Foo>(HttpStatusCode.OK)
        .AddErrorResponse(HttpStatusCode.NotFound)
        .BuildResponseAsync<Foo>(cancellationToken);
}
```

and here's an example usage of `EndpointResponse<TSuccessContent, TErrorContent>`

```csharp
public async Task<EndpointResponse<FooResponse, BadRequestDetails>> UpsertFooAsync(
    FooRequest request,
    CancellationToken cancellationToken)
{
    var requestMessage = requestFactory
        .FromTemplate("/v1/foos")
        .WithBody(request)
        .Build(HttpMethod.Put);

    var response = await httpClient.SendAsync(requestMessage, cancellationToken);
    return await requestFactory
        .FromResponse(response)
        .AddSuccessResponse<FooResponse>(HttpStatusCode.OK)
        .AddSuccessResponse<FooResponse>(HttpStatusCode.Created)
        .AddErrorResponse<BadRequestDetails>(HttpStatusCode.BadRequest)
        .BuildResponseAsync<FooResponse, BadRequestDetails>(cancellationToken);
}
```

### CODE REVIEW

File: src/Atc.Rest.Client/EndpointResponse.cs [hunk 1]
```diff
@@ -1,6 +1,9 @@
+using System;
 using System.Collections.Generic;
 using System.Net;
 
+#pragma warning disable SA1402 // File may only contain a single type
```
Comment (egil) [inline src/Atc.Rest.Client/EndpointResponse.cs:5]: I'm completely fine with disabling that rule globally if you want.

**Review by egil — APPROVED** [review]
LGTM

**Comment by christianhelle** [comment]
I made it as unobtrusive as possible and no existing behavior is changed. I'd like to get this in as I have use case that's waiting for this

File: src/Atc.Rest.Client/EndpointResponse.cs (hunk 1 above)
Comment (christianhelle) [reply src/Atc.Rest.Client/EndpointResponse.cs:5]: Then I will merge this in 😀

Comment (christianhelle) [reply src/Atc.Rest.Client/EndpointResponse.cs:5]: Let's just keep it as a #pragma

**Pull request merged**